object-store/src/main/java/org/hypertrace/config/objectstore/ScopeResolutionIdentifiedObjectStoreWithFilter.java

```diff
@@ -0,0 +1,109 @@
+package org.hypertrace.config.objectstore;
+
+import com.google.protobuf.Descriptors;
+import com.google.protobuf.Message;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.hypertrace.config.service.change.event.api.ConfigChangeEventGenerator;
+import org.hypertrace.config.service.v1.ConfigServiceGrpc;
+import org.hypertrace.core.grpcutils.context.RequestContext;
+
+/**
+ * ScopeResolutionIdentifiedObjectStoreWithFilter is an abstraction over {@link
+ * IdentifiedObjectStoreWithFilter} to allow scope resolution of fetched configs
+ *
+ * @param <T> Config Object
+ * @param <F> Filter
+ * @param <S> Scope
+ */
+public abstract class ScopeResolutionIdentifiedObjectStoreWithFilter<T extends Message, F, S>
+    extends IdentifiedObjectStoreWithFilter<T, F> {
+
+  protected ScopeResolutionIdentifiedObjectStoreWithFilter(
+      ConfigServiceGrpc.ConfigServiceBlockingStub configServiceBlockingStub,
+      String resourceNamespace,
+      String resourceName,
+      ConfigChangeEventGenerator configChangeEventGenerator) {
+    super(configServiceBlockingStub, resourceNamespace, resourceName, configChangeEventGenerator);
+  }
+
+  public List<T> getAllFilteredResolvedConfigs(RequestContext context, F filter, T defaultConfig) {
```

Comment:
Do we need to provide a default? That should likely be built into the store rather than provided on every call.

Reply:
What does it mean to get all resolved? Wouldn't we typically be either getting all unresolved, or getting a single resolved value?

Reply:
> Do we need to provide a default? That should likely be built into the store rather than provided on every call.
Yes agreed - will fix that

> What does it mean to get all resolved? Wouldn't we typically be either getting all unresolved, or getting a single resolved value?
Umm to reduce calls to config-service, we typically ask for all resolved configs - similar to fetch all rules.. example in same - `ai.traceable.anomaly.config.service.detector.anomalydetection.AnomalyDetectionConfigManager`

Reply:
Ah, this is the change that got added for the use case where one of the streaming apps had been fetching/resolving each config per API and instead changed to fetch for all APIs for a given customer right? Makes sense as a use case, but we'd likely want both - getting all resolved (for the config consumer/applier) as well as getting all unresolved (for the config reader/writer ).

Reply:
getting all unresolved is anyways possible using methods from the super-class

Reply:
True, but one thing we have to be careful of is that we're correctly maintaining is-a relationships. For example, if we're resolving a config to produce an object that is a composite of multiple stored objects, it doesn't have its own ID, create time etc. It's a different object shape then the unresolved ones and doesn't support direct deleting or updating. This concern applies at the API level too, so not unique here.

One way would be to add a second type parameter for the resolved vs the persisted objects. Another would be to use a different store that only supports resolving (i.e. doesn't extend existing ones but potentially uses them internally) and the service can use a separate resolved and unresolved store.

Reply:
> umm services which require resolved objects also need unresolved ones 

Not sure I follow this? Do you mean the typed config services? If so, agreed. My point was just that the api to the resolved and unresolved ones is a bit different. In this PR, we add new request arguments to the API, but they share the same response type (driven by the type parameter `T`). For some APIs, the resolved and unresolved objects may both be T, for others, they may be different objects (for example, one without an "id" field). If we take either of the suggestions, they can account for this by not requiring `T` be the same in both cases.

Reply:
if I understand u correctly, what u mean is resolved config type might be different than `T` - I haven't seen such use-cases yet.. in fact the `id` can still be the same as the unresolved config - that's how it would be referenced anyways, especially in case of `getAll`.. having said that, if there is a need of such use-case later, can always add a separate store with different types for resolved and unresolved..
or alternatively can make this more generic as - `ScopeResolutionIdentifiedObjectStoreScopeResolutionIdentifiedObjectStore<T extends Message, U extends Message, F, S>` - though IMO not required for now..

Reply:
> if I understand u correctly, what u mean is resolved config type might be different than T

Correct

> I haven't seen such use-cases yet.. in fact the id can still be the same as the unresolved config - that's how it would be referenced anyways, especially in case of getAll

I would call this a problem. We shouldn't be referencing a resolved config, only evaluating it. If we give it the ID of only one of its component parts, not only is that inaccurate, but that makes it really easy to try to for example upsert the resolved config and break the resolution. Resolved configs should ideally be different messages, even if they're not today, so we'd want to guide new APIs in that direction.

> if there is a need of such use-case later, can always add a separate store with different types for resolved and unresolved..
or alternatively can make this more generic as - ScopeResolutionIdentifiedObjectStoreScopeResolutionIdentifiedObjectStore<T extends Message, U extends Message, F, S> - though IMO not required for now..

So the latter was my second suggestion (and can easily assign T and U to the same thing if needed), the former is similar to my first suggestion, but If we're resolving from a separate store, then that's not a store for writing so I was instead suggesting decoupling the hierarchy rather then extending the existing stores.

Reply:
I still don't see merit in decoupling the hierarchy - it should still extend the existing stores because the written ones and the unresolved ones (which is the input for resolving) would be the same - adding a separate type (U) for resolved one on top of it might make sense..

Reply:
It's two separate sets of APIs. The existing stores provide basic CRUD without any real logic, while the resolution stuff is logic built on top of that basic CRUD to produce new objects (which aren't themselves writable). 

So extending the existing stores and allowing a separate type parameter allows us to put both sets of APIs in the same interface without requiring that they return the same type. This is a bit more convenient for callers since they don't have to manage a separate object (a CRUD store and a resolver - because you'd never use resolution without CRUD as well) but is also pushing the boundaries of/breaking single responsibility principle.